File: erl2c/erl2ocaml_examples/mod03.ml
type unit0 = Unit0

and 'A boxed = Boxed of 'A

and ('A, 'B) either = Left of 'A | Right of 'B

and rgb = R | G | B

and ('A, 'B) pair = Pair of 'A * 'B

and ('A, 'B, 'C) triple = Triple of 'A * 'B * 'C

and 'A list = Cons of 'A * 'A list | Nil

and 'A option = None | Some of 'A

let rec mk_unit'0 = function () -> Unit0

let rec mk_box'1 = function v_A -> Boxed v_A

let rec mk_left'1 = function v_A -> Left v_A

let rec mk_right'1 = function v_B -> Right v_B

let rec zero'2 = function Unit0, v_Val -> v_Val

let rec unbox'1 = function
  | v_Boxed -> ( match v_Boxed with Boxed v_Elem -> v_Elem )

let rec un_either'1 = function
  | v_Either -> (
      match v_Either with Left v_Elem -> v_Elem | Right v_Elem -> v_Elem )

let rec un_pair'1 = function
  | v_Pair -> ( match v_Pair with Pair (v_A, v_B) -> (v_A, v_B) )

let rec first'1 = function Pair (v_F, _) -> v_F

let rec second'1 = function Pair (_, v_S) -> v_S

File: erl2c/erl2ocaml_examples/mod03.erl
-module(mod03).
-compile(export_all).

-enum unit0() :: unit0{}.
-enum boxed(A) :: boxed{A}.
-enum either(A, B) :: left{A} | right {B}.
-enum rgb() :: r{} | g{} | b{}.
-enum pair(A, B) :: pair{A,B}.
-enum triple(A,B,C) :: triple{A,B,C}.
-enum list(A) :: cons{A, list(A)} | nil{}.
-enum option(A) :: none{} | some{A}.

-spec mk_unit() -> unit0().
mk_unit() -> unit0{}.

-spec mk_box(A) -> boxed(A).
mk_box(A) -> boxed{A}.

-spec mk_left(A) -> either(A, _).
mk_left(A) -> left{A}.

-spec mk_right(B) -> either(_, B).
mk_right(B) -> right{B}.
